pkg/controller/stack/stack_config_test.go

```diff
@@ -0,0 +1,219 @@
+package stack
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/stretchr/testify/assert"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+func TestNewStructuredConfigFromJson(t *testing.T) {
+	toJson := func(v any) apiextensionsv1.JSON {
+		b, err := json.Marshal(v)
+		if err != nil {
+			t.Fatal(err)
+			return apiextensionsv1.JSON{}
+		}
+		return apiextensionsv1.JSON{Raw: b}
+	}
+
+	check := func(sourceConfigMap map[string]any, expected []ConfigKeyValue) {
+		structuredConfig, err := NewStructuredConfigFromJSON("", toJson(sourceConfigMap))
+		if assert.NoError(t, err) {
+			configValues := structuredConfig.Flatten()
+			assert.ElementsMatch(t, expected, configValues)
+		}
+	}
+
+	t.Run("We should be able to handle simple key-value pairs", func(t *testing.T) {
+		sourceConfigMap := map[string]any{
+			"aws:region": "us-east-1",
+		}
+		expected := []ConfigKeyValue{
+			{
+				Key: "aws:region",
+				Value: auto.ConfigValue{
+					Value:  "us-east-1",
+					Secret: false,
+				},
+			},
+		}
+		check(sourceConfigMap, expected)
+	})
+	t.Run("We should be able to handle a structured, complex (namespaced) config", func(t *testing.T) {
+		sourceConfigMap := map[string]any{
+			"aws:assumeRole": map[string]any{
+				"roleArn":     "my-role-arn",
+				"sessionName": "my-session-name",
+			},
+			"aws:defaultTags": map[string]any{
+				"tags": map[string]any{
+					"my-tag": "tag-value",
+				},
+			},
+		}
```

Comment:
For what it's worth, I think if you add another key like `"aws:assumeRole.roleArn": "other-role-arn"` this will fail non-deterministically.

Reply:
Could be but, sorry, is it a real problem? For a stack configured like this,

```yaml
aws:assumeRole:
   roleArn: "my-role-arn"
aws:assumeRole.roleArn: "other-role-arn"
```

I wouldn't expect any deterministic behavior here, it would look like a configuration mistake to me.

Reply:
I know it may seem weird, but it's possible so we need to consider it. Hyrum's law.

It would be very helpful to see some more tests around what happens when multiple config maps are provided with overlapping keys like this.